samples/extensions/khr/externalmemory/vulkan_utils.h

```diff
@@ -0,0 +1,389 @@
+        candidate->cl_candidate = cl_candidates[compatible_dev_index];
+    }
+
+    // Check if the device supports the required extensions.
+    if (!check_device_extensions(vk_device, required_device_extensions,
+                                 required_device_extensions_count))
+    {
+        return false;
+    }
+
+    candidate->vk_candidate = vk_device;
+    return true;
+}
+
+// Check if a given OpenCL device supports a particular set of Khronos
+// extensions.
+bool check_khronos_extensions(
+    const cl_device_id cl_device,
+    const char* const* const required_khronos_extensions,
+    const size_t required_khronos_extensions_count)
+{
+    cl_int error = CL_SUCCESS;
+    size_t supported_extensions_count;
+    OCLERROR_RET(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS, 0, NULL,
+                                 &supported_extensions_count),
+                 error, ret);
+    char* supported_extensions =
+        (char*)malloc(supported_extensions_count * sizeof(char));
+    OCLERROR_RET(clGetDeviceInfo(cl_device, CL_DEVICE_EXTENSIONS,
+                                 supported_extensions_count,
+                                 supported_extensions, NULL),
+                 error, err);
+
+    for (size_t i = 0; i < required_khronos_extensions_count; ++i)
+    {
+        if (!strstr(supported_extensions, required_khronos_extensions[i]))
+        {
+            free(supported_extensions);
+            return false;
+        }
+    }
+    free(supported_extensions);
+    return true;
+err:
+    free(supported_extensions);
+ret:
+    return false;
+}
+
+// Find a suitable device for the example, that is, an OpenCL
+// device that is also Vulkan-compatible and that supports the required
+// Vulkan device extensions.
+struct device_candidate
+find_suitable_device(VkInstance instance,
+                     const char* const* required_device_extensions,
+                     const size_t required_device_extensions_count)
+{
+    // Query OpenCL devices available.
+    cl_int error = CL_SUCCESS;
+    bool candidate_found = false;
+    cl_uint platform_count = 0;
+    struct device_candidate found_candidate = {0};
+    OCLERROR_RET(clGetPlatformIDs(0, NULL, &platform_count), error, ret);
+
+    cl_platform_id* platforms =
+        (cl_platform_id*)malloc(platform_count * sizeof(cl_platform_id));
+    OCLERROR_RET(clGetPlatformIDs(platform_count, platforms, NULL), error,
+                 platforms);
+
+    size_t cl_device_count = 0;
+    const char* uuid_khronos_extension[] = {
+        CL_KHR_DEVICE_UUID_EXTENSION_NAME
+    };
+    for (cl_uint platform_id = 0; platform_id < platform_count;
+         ++platform_id)
+    {
+        cl_uint cl_platform_devices_count = 0;
+        OCLERROR_RET(clGetDeviceIDs(platforms[platform_id],
+                                    CL_DEVICE_TYPE_ALL, 0, NULL,
+                                    &cl_platform_devices_count),
+                     error, platforms);
```

Comment:
It would be best not to check for an error here, or at least to filter out the `CL_DEVICE_NOT_FOUND` error.  If a platform supports zero devices (maybe the device is powered down or otherwise disabled) then this should not prevent the sample from executing on a different device.

FYI: https://github.com/KhronosGroup/OpenCL-Docs/issues/1269

Reply:
Changed this! The Platform::getDevices functionality from the C++ headers (which is used in the vulkan_utils.cpp) also follows the logic you described, so this change also fixes consistency between the C and C++ versions of the sample